### ISSUE TITLE
WeeklyPaper5. PyTorch, TensorFlow, Array in Numpy

### ISSUE THREAD

**Issue by finneKIM** [finneKIM/WeeklyPaper#6]
<html>
<body>
<h1>1. 딥러닝 프레임워크인 PyTorch와 TensorFlow를 비교</h1>
<p><strong>PyTorch:</strong></p>
<ul>
<li>PyTorch is a deep learning framework that uses tensors, which are similar to NumPy ndarrays but with added functionalities.</li>
<li>PyTorch tensors can run on GPUs, enabling faster computations for deep learning tasks.</li>
<li>PyTorch provides automatic differentiation through its Autograd module.</li>
<li>PyTorch tensors can be seamlessly converted to and from NumPy arrays, sharing the underlying memory.</li>
<li>The <code>torch.from_numpy()</code> function creates a tensor from a NumPy array. Changes to the NumPy array will reflect in the tensor and vice-versa.</li>
<li>When converting a tensor to a NumPy array, use <code>tensor.detach().cpu().numpy()</code> to avoid gradient issues and ensure the tensor is on the CPU.</li>
</ul>
<p><strong>TensorFlow:</strong></p>
<ul>
<li>TensorFlow is another deep learning framework that also uses tensors as its primary data structure.</li>
<li>TensorFlow tensors are similar to NumPy arrays and PyTorch tensors and can also be run on GPUs.</li>
<li>TensorFlow provides methods to convert between tensors and NumPy arrays.</li>
<li><code>tf.constant()</code> or <code>tf.Variable()</code> can be used to create tensors from Python sequences (like lists or tuples) and NumPy arrays.</li>
<li>Use the <code>.numpy()</code> method to convert a TensorFlow tensor to a NumPy array.</li>
<li>TensorFlow also supports eager execution, which makes debugging easier by evaluating code step-by-step.</li>
<li>TensorFlow provides the <code>TensorArray</code> for dynamic array management.</li>
</ul>

![Image](https://github.com/user-attachments/assets/a8d90a48-ba99-46e2-be9f-782d085e93b5)

항목 | PyTorch | TensorFlow
-- | -- | --
출시연도 및 개발사 | 2017년, Meta AI | 2015년, Google Brain arxiv.org+5builtin.com+5en.wikipedia.org+5en.wikipedia.org+1fr.wikipedia.org+1
그래프 실행 방식 | 동적(dynamic) 실행 – Pythonic하고 직관적인 디버깅 가능 | 정적(static) 그래프 기반 (TF1.x) → TF2.x부터는 Eager 실행도 지원
코딩 스타일 | Python 친화적, Imperative | Keras 스타일 레벨 API + 그래프 중심
시각화 도구 | Visdom 사용 (기능 제한적) | TensorBoard: 풍부한 시각화 및 프로파일링 제공
분산 훈련 지원 | Python 비동기 실행 덕분에 상대적으로 간단 | 분산 설정 시 직접 코드 작성 필요하지만, 향상된 전략 API 존재
프로덕션 배포 | Flask/Django 사용 필요, 직접 서버 구성 | TensorFlow Serving, 모바일/임베디드 배포 (TFLite) 등 ★배포 친화적★
커뮤니티 & 문서 | 활발한 연구 커뮤니티 지원, 적용 예제 풍부 | 기업용 프로덕션 중심 사용, 공식 문서 및 엔터프라이즈 사례 다수
장단점 요약 | **장점**: Pythonic, 동적 그래프, 빠른 프로토타이핑. **단점**: 시각화 도구 및 배포 인프라 부족 | **장점**: 강력한 시각화, 배포 생태계, 모바일 지원. **단점**: 초기 학습 난이도 높고, static 그래프 디버깅이 복잡 |   |  

<h3>🧭 선택 가이드</h3>
<ul>
<li><strong>연구 또는 빠른 실험/프로토타이핑</strong>에 초점을 둔다면 → <strong>PyTorch</strong></li>
<li><strong>대규모 시스템 구축, 배포, 모바일/엣지 기기 지원</strong>이 중요하다면 → <strong>TensorFlow</strong></li>
</ul>
<hr>
<h3>참고</h3>
<p><a href="https://builtin.com/data-science/pytorch-vs-tensorflow">https://builtin.com/data-science/pytorch-vs-tensorflow</a></p>
<p><a href="https://www.geeksforgeeks.org/python/difference-between-pytorch-and-tensorflow/">https://www.geeksforgeeks.org/python/difference-between-pytorch-and-tensorflow/</a></p>

**Comment by finneKIM** [comment]
<html>
<body>
<h1>2. PyTorch에서 텐서란 무엇이고, NumPy의 Array(배열)과의 차이점은?</h1>
<p><strong>NumPy Arrays</strong></p>
<ul>
<li>NumPy is a fundamental Python library for numerical computing.</li>
<li>It introduces the <code>ndarray</code> object, which is a multi-dimensional array optimized for numerical operations.</li>
<li>NumPy arrays are faster and more memory-efficient than Python lists for numerical tasks.</li>
<li>It provides a wide range of functions for array manipulation, mathematical operations, and linear algebra.</li>
<li>NumPy arrays are primarily CPU-based and do not offer built-in GPU acceleration.</li>
</ul>
<p><strong>PyTorch Tensors</strong></p>
<ul>
<li>PyTorch is a deep learning framework that uses tensors as its primary data structure.</li>
<li>Tensors are similar to NumPy's <code>ndarray</code> but with added capabilities for GPU acceleration and automatic differentiation.</li>
<li>PyTorch tensors can be created using <code>torch.tensor()</code> or converted from NumPy arrays.</li>
<li>PyTorch provides a mechanism for converting tensors to NumPy arrays using the <code>.numpy()</code> method.</li>
<li>PyTorch tensors can be moved to and processed on GPUs, significantly speeding up computations for machine learning tasks.</li>
</ul>
<hr>
<h2>주요한 차이점과 병합</h2>
<ul>
<li>
<p><strong>Automatic Differentiation:</strong></p>
<p>PyTorch tensors have built-in support for automatic differentiation, while NumPy arrays do not. TensorFlow also has built-in support for automatic differentiation.</p>
</li>
<li>
<p><strong><code>GPU Support:</code></strong></p>
<p><strong>PyTorch tensors and TensorFlow tensors can leverage GPUs for accelerated computation</strong>, while NumPy arrays have limited GPU support.</p>
</li>
<li>
<p><strong>Conversion:</strong></p>
<p>Both PyTorch and TensorFlow have mechanisms for easy conversion between their tensors and NumPy arrays.</p>
</li>
<li>
<p><strong>Use Cases:</strong></p>
<p>NumPy is primarily for numerical and scientific computing, <strong>while PyTorch and TensorFlow are designed for deep learning applications.</strong></p>
</li>
<li>
<p><strong>Memory Sharing:</strong></p>
<p><strong>PyTorch tensors and NumPy arrays share memory</strong>, which means changes in one will affect the other, unless a copy is made.</p>
</li>
</ul>
<hr>
<h1>Tensors and NumPy Array: Key-Differences</h1>

Feature | PyTorch Tensors | NumPy Arrays
-- | -- | --
Definition | Tensors are multi-dimensional arrays with uniform data types with more optimization for Deep Learning | They are also multi-dimensional arrays with a uniform data type with less support for Deep Learning.
Syntax and Interface | You can use the torch.tensor() method to create the Tensors. | To create the NumPy Arrays, the np.array() method is used.
Automatic Differentiation | It supports the Built-in automatic differentiation using PyTorch’s Autograd module. | There is no support for automatic differentiation.
GPU Support | We can integrate it with CUDA-enabled GPUs for accelerated computation. | It provides Limited support for GPU. Thus, we need additional libraries for GPU.
Dynamic Computational Graph | It supports dynamic computation graphs in which the graph can be changed at the run time. | It supports the Static computation graph in which the computation graph is defined before execution.
Performance | It supports efficient GPU acceleration for deep learning tasks. | It is efficient for general-purpose numerical computations but less optimized for deep learning.
Deployment | It supports the deployment learning models in production environments. | We require additional steps for deployment and integration with deep learning frameworks.
Memory Management | It has Automatic memory management with garbage collection. | It has Manual memory management. Thus, we need to implement explicit memory deallocation.
Integration with Deep Learning Frameworks | It supports Native integration with PyTorch’s deep learning ecosystem for seamless model development. | It also requires additional steps for integration with deep learning frameworks like TensorFlow or Keras.
Parallelization | It supports parallel operations across multiple CPU or GPU cores. | The Parallel operations depend on the underlying linear algebra libraries like BLAS and CPU/GPU hardware.

![Image](https://github.com/user-attachments/assets/ab8999b6-c8c8-4e88-b547-254056eb4953)

<hr>
<h3>참고</h3>
<p><a href="https://www.geeksforgeeks.org/deep-learning/pytorch-tensor-vs-numpy-array/">https://www.geeksforgeeks.org/deep-learning/pytorch-tensor-vs-numpy-array/</a></p>
<p><a href="https://www.kaggle.com/code/amirmotefaker/pytorch-vs-numpy">https://www.kaggle.com/code/amirmotefaker/pytorch-vs-numpy</a></p>
<p><a href="https://www.reddit.com/r/learnmachinelearning/comments/12u9sn1/tensors_vs_numpy_arrays/">https://www.reddit.com/r/learnmachinelearning/comments/12u9sn1/tensors_vs_numpy_arrays/</a></p>

</body>
</html>

**Comment by finneKIM** [comment]
<html>
<body>
<h3>Garbage Collection에 관한 추가 공부 (Deployment 관련)</h3>

항목 | 내용 요약
-- | --
가비지 컬렉션(GC)이란? | JVM이 Heap 영역에서 더 이상 참조되지 않는 불필요한 메모리를 자동으로 회수하는 기능으로, 자바 등에서 메모리 누수 방지 및 안정성을 보장함 danyoujeong.tistory.com
가비지(Garbage)의 정의 | Stack에서 도달 불가능한(Unreachable) 객체를 의미하며, 이런 객체는 GC 대상이 됨. Stop-the-world(STW)는 GC 시 애플리케이션 실행이 중단되는 현상
GC 기본 동작 방식 (Mark & Sweep) | 1. Mark: Root 객체부터 접근 가능한 객체를 마킹2. Sweep: 마킹되지 않은(참조되지 않은) 객체 제거3. Compact: 힙 단편화를 줄이기 위해 메모리 재배치(선택적으로 수행)
세대별 GC 구조 | JVM Heap은 Young (Eden + Survivor S0/S1)와 Old 영역으로 나뉘며, 이는 Weak Generational Hypothesis (짧게 사용→Young, 오래 사용→Old) 기반
Minor GC (Young 영역) | Eden이 가득 차면 발생, 주로 짧게 실행되며 Stop-the-world 시간을 최소화. 살아남은 객체는 Survivor로 이동하고 age 증가
Major (Full) GC (Old 영역) | Old 영역이 포화되면 실행. Heap 전체를 검사하며 시간이 오래 걸림
Promotion 과정 | 1. 객체 생성 → Eden2. Eden 포화 → Minor GC 실행 → Sweep/Mark3. 살아남은 객체는 S0/S1로 이동 → age 부여4. 일정 age 이상 → Old로 승격(Promotion)5. Old 영역이 꽉 차면 Major GC 발생
예상 면접 질문 요약 | - 가비지 컬렉터란? GC 기능 수행 주체, 자동 메모리 회수. 수동 언어(C/C++) 대비 차이 설명 가능- 동작 방식? Young/Old 세대 구조, Mark-Sweep 알고리즘, Stop‑the‑world 개념, Minor/Major GC 단계


<h3>📌 요약</h3>
<ul>
<li>GC는 자바의 자동 메모리 회수 시스템으로 메모리 누수를 방지.</li>
<li>“Mark &amp; Sweep”으로 필요 없는 메모리를 탐색·삭제하며, 필요 시 힙 재정렬(Compact).</li>
<li>세대별 영역(Young vs Old) 구조를 활용해 Minor GC (짧게, 빈번하게)와 Major GC (길고 드물게)를 수행.</li>
<li>살아남는 객체는 Survivor 단계를 거치며 최종적으로 Long‑lived 객체는 Old 영역으로 승격.</li>
<li>면접 대비 시에는 GC의 필요성, STW의 영향, 세대별 구조와 수행 방식 등을 명확히 설명하는 것이 중요.</li>


------
<html>
<body>
<h3>✅ 가비지 컬렉션(Garbage Collection) 개념 정리 (주석 포함)</h3>

항목 | 설명 | 주석
-- | -- | --
Garbage Collection (GC) | 프로그램 실행 중 더 이상 사용하지 않는 메모리를 자동으로 정리해주는 시스템 | // 메모리를 쓰고 나서 안 쓰게 된 데이터를 자동으로 청소하는 시스템
Heap 영역 | 객체가 저장되는 메모리 공간 | // new 키워드로 만든 데이터들이 저장되는 큰 저장소
Stack 영역 | 메서드 호출 시 사용하는 임시 메모리 공간 | // 함수나 변수들이 잠깐 저장되는 공간
Garbage (가비지) | 더 이상 참조되지 않아 쓸모없는 객체 | // 아무도 사용하지 않는 데이터 찌꺼기
Unreachable (도달 불가능한) | 어떤 변수도 더 이상 참조하지 않는 상태 | // 다른 코드에서 접근할 수 없는 상태
Stop-the-world (STW) | GC를 수행하는 동안 프로그램 실행을 멈추는 현상 | // 메모리를 정리하는 동안 전체 프로그램이 일시정지되는 현상
Mark | 접근 가능한 객체를 표시 | // 살아있는 데이터에 표시를 붙임
Sweep | 표시되지 않은 객체를 제거 | // 표시 안 된 찌꺼기 데이터를 청소
Compact | 메모리 조각을 정리해 빈틈없이 정렬 | // 데이터 사이사이 빈 공간을 없애기 위해 정리
Generational GC (세대별 GC) | 객체의 생존 기간에 따라 메모리를 나눠 관리 | // 새로 만든 데이터와 오래된 데이터를 따로 관리함
Young 영역 | 새로 생성된 객체들이 저장되는 영역 | // 방금 만든 데이터들이 저장되는 공간 (Eden + Survivor)
Old 영역 | 오래 살아남은 객체들이 저장되는 영역 | // 한동안 계속 쓰이는 데이터들이 모이는 공간
Eden | 객체가 처음 생성되는 공간 | // 데이터가 처음 만들어질 때 저장되는 자리
Survivor S0/S1 | Eden에서 살아남은 객체가 이동하는 공간 | // 1차 생존 테스트를 통과한 데이터들이 잠깐 머무는 공간
Minor GC | Young 영역에서 수행되는 GC | // 새 데이터들만 빠르게 정리하는 과정
Major (Full) GC | Heap 전체를 대상으로 하는 GC | // 모든 데이터를 대상으로 오래 걸리는 대청소
Promotion (승격) | 여러 번 살아남은 객체를 Old 영역으로 이동 | // 여러 번 테스트 통과한 데이터를 오래된 공간으로 이동
Age | 객체가 살아남은 GC 횟수 | // 데이터가 몇 번이나 정리에서 살아남았는지를 나타냄
Root 객체 | 프로그램에서 참조를 시작하는 객체 | // 기준이 되는, 다른 객체를 연결해주는 시작점


<hr>
<h3>📌 요약 예시 문장</h3>
<blockquote>
<p>자바에서 GC는 사용하지 않는 메모리를 자동으로 정리해주는 시스템입니다. 메모리는 Heap 영역에 저장되고, 사용되지 않게 되면 Garbage로 분류되어 Mark–Sweep 방식으로 제거됩니다. 객체의 생존 기간에 따라 Young과 Old 영역으로 구분되며, 각각 Minor GC와 Major GC를 통해 관리됩니다.</p>
</blockquote>

</body>
</html>


--------
### 참고

https://danyoujeong.tistory.com/214
